util/test/demos/vk/vk_ray_query.cpp

```diff
@@ -0,0 +1,591 @@
+                              VK_BUFFER_USAGE_ACCELERATION_STRUCTURE_STORAGE_BIT_KHR |
+                                  VK_BUFFER_USAGE_SHADER_DEVICE_ADDRESS_BIT_KHR),
+        VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT);
+
+    VkAccelerationStructureCreateInfoKHR tlasCreateInfo = {
+        VK_STRUCTURE_TYPE_ACCELERATION_STRUCTURE_CREATE_INFO_KHR};
+    tlasCreateInfo.buffer = tlasBuffer.buffer;
+    tlasCreateInfo.size = tlasBuildSizesInfo.accelerationStructureSize;
+    tlasCreateInfo.type = VK_ACCELERATION_STRUCTURE_TYPE_TOP_LEVEL_KHR;
+
+    VkAccelerationStructureKHR tlas;
+    CHECK_VKR(vkCreateAccelerationStructureKHR(device, &tlasCreateInfo, VK_NULL_HANDLE, &tlas));
+
+    AddressedBuffer tlasScratchBuffer(
+        this,
+        vkh::BufferCreateInfo(
+            tlasBuildSizesInfo.buildScratchSize,
+            VK_BUFFER_USAGE_STORAGE_BUFFER_BIT | VK_BUFFER_USAGE_SHADER_DEVICE_ADDRESS_BIT_KHR),
+        VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT);
+
+    tlasBuildGeometryInfo.scratchData.deviceAddress = tlasScratchBuffer.address;
+    tlasBuildGeometryInfo.dstAccelerationStructure = tlas;
+
+    asBuildRangeInfosVector[0].primitiveCount = 1;
+
+    {
+      VkCommandBuffer cmd = GetCommandBuffer();
+      CHECK_VKR(vkBeginCommandBuffer(cmd, vkh::CommandBufferBeginInfo()));
+      vkCmdBuildAccelerationStructuresKHR(cmd, 1, &tlasBuildGeometryInfo, &asBuildRangeInfos);
+      CHECK_VKR(vkEndCommandBuffer(cmd));
+      Submit(99, 99, {cmd});
+    }
+
+    /*
+     * Create a pointless copy for more API coverage
+     */
+    AddressedBuffer newBlasBuffer(
+        this,
+        vkh::BufferCreateInfo(blasBuildSizesInfo.accelerationStructureSize,
+                              VK_BUFFER_USAGE_ACCELERATION_STRUCTURE_STORAGE_BIT_KHR |
+                                  VK_BUFFER_USAGE_SHADER_DEVICE_ADDRESS_BIT_KHR),
+        VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT);
+
+    VkAccelerationStructureCreateInfoKHR newBlasCreateInfo = {
+        VK_STRUCTURE_TYPE_ACCELERATION_STRUCTURE_CREATE_INFO_KHR};
+    newBlasCreateInfo.buffer = newBlasBuffer.buffer;
+    newBlasCreateInfo.size = blasBuildSizesInfo.accelerationStructureSize;
+    newBlasCreateInfo.type = VK_ACCELERATION_STRUCTURE_TYPE_BOTTOM_LEVEL_KHR;
+
+    VkAccelerationStructureKHR newBlas;
+    CHECK_VKR(vkCreateAccelerationStructureKHR(device, &newBlasCreateInfo, VK_NULL_HANDLE, &newBlas))
+
+    /*
+     * Create uniforms
+     */
+    constexpr VkBufferUsageFlags uniformBufferUsageFlags =
+        VK_BUFFER_USAGE_ACCELERATION_STRUCTURE_BUILD_INPUT_READ_ONLY_BIT_KHR |
+        VK_BUFFER_USAGE_SHADER_DEVICE_ADDRESS_BIT_KHR | VK_BUFFER_USAGE_STORAGE_BUFFER_BIT;
+
+    AddressedBuffer uniformVertexBuffer(
+        this,
+        vkh::BufferCreateInfo(vertexBufferSize,
+                              uniformBufferUsageFlags | VK_BUFFER_USAGE_VERTEX_BUFFER_BIT),
+        VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT);
+    AddressedBuffer uniformIndexBuffer(
```

Comment:
These are named `uniform` but they look to be normal VBs/IBs? They also don't seem to need device addresses, they could be normal buffers.